compiler-core/src/language_server/code_action.rs

```diff
@@ -7562,3 +7562,374 @@ impl<'ast> ast::visit::Visit<'ast> for RemovePrivateOpaque<'ast> {
+    fn visit_typed_clause(&mut self, clause: &'ast ast::TypedClause) {
+        let ast::TypedClause {
+            pattern,
+            alternative_patterns,
+            then,
+            location,
+            guard,
+        } = clause;
+
+        // We can only flatten a case expression in a very specific case:
+        // - This pattern may be introducing multiple variables,
+        // - The expression following this branch must be a case, and
+        // - It must be matching on one of those variables
+        //
+        // For example:
+        //
+        // ```gleam
+        // Wibble(a, b, 1) -> case a { ... }
+        // Wibble(a, b, 1) -> case b { ... }
+        // ```
+
+        // We can't have any alternative patterns...
+        if alternative_patterns.is_empty()
+            // ...and the then clause must be a single case expression...
+            && let Some(TypedExpr::Case {
+                subjects, clauses, ..
+            }) = then.expression_flattening_block()
+            // ...that is matching a single variable...
+            && let [TypedExpr::Var { name, .. }] = subjects.as_slice()
+            // ...and that variable must be one the variables we brought into
+            // scope in this branch.
+            && let Some(variable) = pattern
+                .iter()
+                .flat_map(|pattern| pattern.bound_variables())
+                .find(|variable| variable.name() == *name)
+        {
+            // There's one last condition to trigger the code action: we must
+            // actually be with the cursor over the pattern or the nested case
+            // expression!
+            //
+            // ```gleam
+            // case wibble {
+            //   Ok(a) -> case a {
+            // //^^^^^^^^^^^^^^^ Anywhere over here!
+            //   }
+            // }
+            // ```
+            //
+            let pattern_location = pattern
+                .first()
+                .map_or(*location, |first| first.location())
+                .merge(&pattern.last().map_or(*location, |end| end.location()));
+
+            let valid_location = match subjects.last() {
+                Some(last_subject) => pattern_location.merge(&last_subject.location()),
+                None => pattern_location,
+            };
+            let valid_range = self.edits.src_span_to_lsp_range(valid_location);
+
+            if within(self.params.range, valid_range) {
+                self.collapsed = Some(Collapsed {
+                    outer_clause_span: *location,
+                    outer_guard: guard,
+                    matched_variable: variable,
+                    matched_pattern_span: pattern_location,
+                    inner_clauses: clauses,
+                });
+                // We're done, there's no need to keep exploring as we know the cursor
+                // is over this pattern and it can't be over any other one!
+                return;
```

Comment:
This means that nested further cases don't get checked, even though they could also be collapsed.

Reply:
ah! I misunderstood the code before. I feel vindicated on my request to make that helper function now 😁